stdlib/src/builtin/string.mojo

```diff
@@ -1333,36 +1333,93 @@ struct String(
             substr._strref_dangerous(), start=start
         )
 
-    fn split(self, delimiter: String) raises -> List[String]:
+    @always_inline
+    fn split(
+        self, delimiter: StringRef, maxsplit: Int = Int.MAX
+    ) raises -> List[String]:
         """Split the string by a delimiter.
 
         Args:
           delimiter: The string to split on.
+          maxsplit: The maximum number of splits to allow.
 
         Returns:
           A List of Strings containing the input split by the delimiter.
 
         Raises:
           Error if an empty delimiter is specified.
         """
-        if not delimiter:
+        return self.split(str(delimiter), maxsplit)
+
+    fn split(
+        self, delimiter: Optional[String] = None, maxsplit: Int = Int.MAX
```

Comment:
Instead of having `delimiter` taking an `Optional`, we should probably use an overload.

Reply:
What would the practical advantage be of doing it that way you think?

Reply:
Move the implementation in the two branches to overloads.
```moojo
fn split(self, maxsplit: Int = ...)
fn split(self, delimiter: String, maxsplit: Int = ...)
```

Reply:
Hmmm I wasn't convinced at first but on second thought I think you're right

Reply:
I would prefer fewer overloads in this case. Firstly, it wouldn't be necessary to define the default value for `maxsplit` twice. Secondly, a call to `some_string.split(bar)` could become ambiguous if `bar` is implicitly convertible to both `String` and `Int`. Given that we cannot make these arguments keyword-only (because the Python API dictates that these are positional-or-keyword passable), a single overload is the safest way.

Reply:
Yea, I think it's actually the other way around: it doesn't make much sense to have this function taking an optional value.

Reply:
I think it makes sense to have it take an Optional since `.split(None)` is completely valid in Python.

Also, `maxsplit` should default to `-1`. (since `.split(maxsplit=-1)` is the same as doing `.split()`)